pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go

```diff
@@ -44,12 +46,52 @@ func (gce *GCECloud) ensureInternalLoadBalancer(clusterName, clusterID string, s
 	ports, protocol := getPortsAndProtocol(svc.Spec.Ports)
 	scheme := schemeInternal
 	loadBalancerName := cloudprovider.GetLoadBalancerName(svc)
-	specialHCPath, specialHCPort := v1_service.GetServiceHealthCheckPathPort(svc)
-	shared := specialHCPath == "" && specialHCPort == 0
-	backendServiceName := getBackendServiceName(loadBalancerName, clusterID, shared, scheme, protocol)
-	backendServiceLink := gce.getBSLink(backendServiceName)
+	shared := !v1_service.RequestsOnlyLocalTraffic(svc)
+	backendServiceName := makeBackendServiceName(loadBalancerName, clusterID, shared, scheme, protocol)
+	backendServiceLink := gce.getBackendServiceLink(backendServiceName)
 
-	wantedFwdRule := &compute.ForwardingRule{
+	// Ensure instance groups exist and nodes are assigned to groups
+	igName := makeInstanceGroupName(clusterID)
+	igLinks, err := gce.ensureInstanceGroups(igName, nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Lock
+	if shared {
+		gce.sharedResourceLock.Lock()
+		defer gce.sharedResourceLock.Unlock()
+	}
```

Comment:
If we always attempt to delete the inverse resource, I guess this lock would be fairly broad...For now in ELB we only lock during `createNodesHealthCheck -> createTargetPool`  and `deleteNodesHealthCheck`.